Perf protocol implementation

Ralith opened the pull request:
For https://github.com/quicwg/base-drafts/wiki/Performance-Hackathon

Matthias247 commented on perf/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+rcgen = "0.8"
+rustls = "0.19"
+structopt = "0.3"
+tokio = { version = "1.0.1", features = ["rt-multi-thread", "macros"] }
```
I think this should use a single-threaded runtime given that `bulk` shows very clearly that multithreaded performs worse

Ralith replied:
I was thinking about the many-connection case, but sure, might as well play to our strengths.

Matthias247 commented on perf/src/main.rs:
```diff
@@ -0,0 +1,118 @@
+        .unwrap();
+    server_config.protocols(&[b"perf"]);
+
+    let server_config = server_config.build();
```
I guess this is ok for starting the thing. However I think this should later adjust settings for increasing perf. E.g. flow control limits might be raised. Limits should probably equal to what `bulk` does.

Matthias247 commented on perf/src/main.rs:
```diff
@@ -0,0 +1,118 @@
+use tracing::{debug, error, info};
+
+#[tokio::main]
+async fn main() {
```
This always starts the server. I think quinn should also contain a client. Should this be an extra executable? Or maybe make it possible to configure via CLI (`perf --client`, `perf --listen ip`) which peer is used.

Matthias247 commented on perf/src/main.rs:
```diff
@@ -0,0 +1,118 @@
+        let stream = stream?;
+        let connection = connection.clone();
+        tokio::spawn(async move {
+            if let Err(e) = accept_uni(connection, stream).await {
```
I would rather call it `handle_uni`. `accept` sounds like it would wait for a stream. However that already had happened. Same for `accept_bi`

Matthias247 commented on perf/src/main.rs:
```diff
@@ -0,0 +1,118 @@
+        .await
+        .context("reading request")?;
+    let n = u64::from_be_bytes(buf);
+    debug!("got req for {} bytes", n);
```
This actually misses draining the remaining request body, which is important for upload tests.

I would recommend to copy&paste https://github.com/quinn-rs/quinn/blob/main/bench/src/bulk.rs#L88-L101.
Optionally a `async fn drain()` which does the same could also be added to `RecvStream`.

Matthias247 commented on perf/src/client.rs:
```diff
@@ -0,0 +1,183 @@
+    cfg.protocols(&[b"perf"]);
+    let mut cfg = cfg.build();
+    let tls_cfg: &mut rustls::ClientConfig = Arc::get_mut(&mut cfg.crypto).unwrap();
+    tls_cfg
```
Maybe move this behind an `--insecure` flag? There can definitely be use cases to run this against places with a valid cert. For those use-cases it might also be nice to either
- specify the ip independent from the hostname (similar to how `curl --resolve` works)
- allow to overwrite the SNI

I would recommend the first option. But this is also fine for a follow-up, and i'm happy to add it.

Matthias247 commented on perf/src/client.rs:
```diff
@@ -0,0 +1,183 @@
+    }
+}
+
+async fn request_uni(mut send: quinn::SendStream, size: u64) -> Result<()> {
```
This as well as `request_bi` will require another parameter to specify how much data to upload (and an implementation of it).

Matthias247 commented on perf/src/server.rs:
```diff
@@ -0,0 +1,160 @@
+
+async fn handle_bi(send: quinn::SendStream, recv: quinn::RecvStream) -> Result<()> {
+    let bytes = read_req(recv).await?;
+    respond(bytes, send).await?;
```
draining of the request stream is still missing.  What I'm not sure is whether this actually should happen concurrently to sending a request. I implemented it now somewhere else as:

```
let bytes = read_header(&mut recv).await?;

let drain_request_fut = drain_data(recv);
let send_response_fut = send_response(bytes, send);

let ((), ()) = tokio::try_join!(drain_request_fut, send_response_fut)?;

Ok(())
```

Matthias247 replied:
Reading through the update https://github.com/nibanks/quic-perf/pull/2/files , concurrent sending of the response doesn't seem required and the `try_join!` can be dropped.

Matthias247 commented on perf/src/client.rs:
```diff
@@ -0,0 +1,183 @@
+
+    endpoint.wait_idle().await;
+
+    // TODO: Print stats
```
Yes, we will need those to make it useful. Probably similar to what exists in `bulk`

Ralith commented:
> Should we open an issue for the QUINN BUG?

Whoops, that was fixed in #1032. Removed comment

> I don't think keeping these commits separate is super useful

Squashed.
